Use Promises in asynchronous actions, replace suspend/capture with pause/resume

istateside opened the pull request:
We could not use Promises in Capturama reliably before, because Capturama did not handle Promises correctly in its event loop. With https://github.com/movableink/capturama/pull/90 , that is no longer a problem. We can move away from the callback-y nature of CD.get, and start using Promises.

This PR does a few things:
- Sets up a very basic webpack config to bundle together cropduster and our new dependency
- Changes CD.get and its related functions to create and return a Promise that resolves when the network request finishes, and rejects if an error is hit.
- Stops outputting a separate ES5 version of cropduster. This version of cropduster is intended to be used as a dependency of other apps, and there's no current webpack set up to compile it down to ES5. We could add babel in pretty easily.
- Decreases the file size of `dist/cropduster.js` from 7Kb to ~5Kb by uglifying the output.

Things this PR does not do:
- Breaks the existing API. The existing API for CD.get is to pass it a callback. There is no return value. That will continue to work. If no callback is passed into the app, it sets the default callback as an empty function, and continues on its merry way
- Deprecates the existing API. We could console log a message inside CD.get to communicate that users should be treating CD.get like a Promise, if we want to move people towards the new API.
- Use Promises to their full extent- we could rewrite CD.getImages to use Promises more efficiently, but it would likely change that API completely.

istateside commented:
https://app.clubhouse.io/movableink/story/11388/use-promises-in-the-data-sources-app

spencer516 commented on README.md:
```diff
@@ -162,6 +162,11 @@ console.log('If user clicks on the web crop, they will go to http://example.com'
 
 ## Changelog
 
+### 4.2.0
```
Is this worth a major version bump now that we're introducing promises?

spencer516 commented on src/cropduster.js:
```diff
@@ -201,33 +203,53 @@ const CD = {
     var msg = "xhr: " + url;
 
-    var req = new XMLHttpRequest();
+    return new Promise(function(resolve, reject) {
```
Can we split this method into two — where the behavior of `get` is the same and introduce a new one called `ajax` that returns the promise-only version? (This way we could start deprecating the callback-y form of this). 

The `get` method would end up looking something like:

```js
get(url, options, callback) {
  this.ajax(url, options).then(
    ({ contentType, status, data }) => {
      callback(data, status, contentType);
    }, 
    ({ status }) => {
      callback(null, status);
    }
  ); 
}
```

(Would we want to call it `ajax`? Or, `getPromise`?)

mnutt replied:
I think we'd want to make `get` the primary function--`getCors` uses `get` so you'd end up with `getCorsPromise` or something and it feels a bit weird to have `promise` in function names.  Wouldn't we just be able to do:

```
if(callback) {
  console.log('callbacks are deprecated, prefer promises');
  callback(...);
}
```

Inside the `onerror` / `onsuccess` callbacks?

spencer516 commented on src/cropduster.js:
```diff
@@ -238,17 +260,24 @@ const CD = {
-    img.src = url;
-    CD.suspend(options.maxSuspension, msg);
+
+    return new Promise(function(resolve, reject) {
```
I guess the same feedback applies to `getImage` and `getImages` too...can there be separate methods for the promise-based implementations? 

It also looks like we do some complex argument stuff so there's the optional `options` argument — but it looks we only use `maxSuspension` on it. We could change the `getImagePromise` signature to be more explicit:

Something like:

```js
getImage(...args) {
  callback = args.pop();
  url = args[0];
  options = args[1] || {};

  this.getImagePromise(url, options.maxSuspension).then(
    img => callback(img),
    _ => callback(null)
  );
}
```

spencer516 commented on src/cropduster.js:
```diff
@@ -265,48 +294,51 @@ const CD = {
 
-    for(var i = 0; i < urls.length; i++) {
-      (function(url, i){
+    return new Promise(function(resolve, reject) {
```
With the `getImagePromise` method, we could also rewrite `getImages` to leverage that directly and simplify quite a bit (this is doing a lot of manual bookkeeping!).

mnutt commented:
I think it's actually up for debate if we want to even offer the webpacked version going forward. Most of the cropduster usage I've seen that _hasn't_ been included via npm has been for undesirable use cases like pasting into "Injected Javascript" and such.

mnutt commented:
Before we release a new cropduster version we may want to just fork-and-rename Promiscuous so that we can get it in NPM and aren't relying on a git sha.

spencer516 commented:
@mnutt — any reason why we need to use our own version of Promiscuous at all and not just depend on the already-published version? (It looks like our fork of it is even with the main version).

mnutt commented:
It's actually depending on a branch: 

https://github.com/movableink/promiscuous/tree/mn/capturama-compatible

I don't think it would make sense to try to get those changes merged in since they have no real value except to us.

spencer516 commented:
It _is_ depending on a branch? Or it's meant to depend on a branch? 

The resolved SHA in the `yarn.lock` file  is pointing to the master branch:

https://github.com/movableink/cropduster/pull/20/files#diff-8ee2343978836a779dc9f8d6b794c3b2R2731

mnutt commented:
Ahh, it's _meant_ to depend on a branch. Probably won't work in crops if it doesn't.

istateside commented:
Huh. Weird. It was working in crops, I probably didn't hit the edge case that it's supposed to account for. I'll point to the branch.

istateside commented:
Updated this branch to include the promiscuous fork as a file int he `src`directory instead of as a dependency.

Including it as a dependency introduced a circular dependency (Promiscuous depends on CD so it can call CD.suspend in its finalize function, but CD depends on Promiscuous so it can use its Promise object)

I changed the Promiscuous implementation a bit, so that it returns a factory function, which accepts a suspend and capture function. It uses those functions in the place of where we were calling CD.suspend or CD.capture (although we pass those functions as the suspend/capture params). When called, the factory returns a Promise implementation using those functions (identical to what the fork was doing before).

It's a bit messy, but it lets us avoid the circular dependency. With this change, we could move Promiscuous back to an external dependency (pointing to this updated version of our fork), or we could just leave it here. /shrug

mnutt commented on src/cropduster.js:
```diff
@@ -195,118 +203,149 @@ const CD = {
     url = args[0];
-    callback = args.pop();
-    options = args[1] || {};
+    if (typeof options === 'function') {
```
I haven't run this code yet, but just looking at it, it looks like we set `options = args[1] || {}`, which sets it to the function. It'll probably _work_ but I think the intention is to set `callback = options; options = {}` since the user didn't pass actual options?

mnutt commented:
I also think that, given that we have the opportunity to change the API in the Promise case, we should maybe try to adhere to the `fetch` API, as soon enough `get` can be mostly replaced by a wrapper around `fetch` anyway.

istateside commented:
Updated this PR to remove the Promiscuous dependency, and set up a Capturama PR that injects Promiscuous here: https://github.com/movableink/capturama/pull/90

Looking into using the fetch API, although I might put that off for another card if the API needs to change significantly from what's in place here

istateside commented:
Alrighty:

Updated this PR to match the changes made in this Capturama update: https://github.com/movableink/capturama/pull/90

- I've bumped the version to 5.0
- Removed CD.suspend and CD.capture entirely, and replaced them with CD.pause and CD.resume, which each do little besides calling MICapture.pause and MICapture.resume. The bookkeeping for open suspends that was done in suspend/capture before is now handled directly in Capturama, so CD.pause and resume are very simple. MICapture.begin and MICapture.end both still exist and operate identically to how they did before, so any existing apps that use older versions of cropduster will continue to work. 
- get, getImages, getCORS all work with promises now.
- I've updated the webpack config to output two versions of the compiled cropduster- one for use in browsers and one as an import in other projects.
- the tests are cleaned up a bit, and some tests were changed or added for the new features.

Working on getting the Travis config working with google chrome instead of phantom js, and then verifying the two outputs from webpack work correctly, and this should be good to go.

istateside commented:
Didn't see michael's original comment about the get/getPromise stuff- I agree that carrying around the extra methods feels awkward- gonna rewrite it to deprecate callbacks and work with the existing API

istateside commented:
Updated this PR to get rid of the extra methods and use Promises where possible.

The only real breaking change in this update is in CD.getImages. Before, you'd pass two callbacks, `afterAll`, then `afterEach`.

Since `afterAll` should now be handled by promise chains like so...
```javascript
CD.getImages(arrayOfImages).then((images) => {
  // do something with the images
});
```

... it doesn't makes sense for afterAll to be the first argument any more. I've switched the order of afterAll (which will log a deprecation message if it's passed, but will still be called) and afterEach (which can still be used as before, and will be called after each individual image promise resolves).

The follow up PR to implement the fetch API instead of XMLHttpRequest is [here](https://github.com/movableink/cropduster/pull/24), but is dependent on another Capturama change [here](https://github.com/movableink/capturama/pull/91/files)

mnutt commented on README.md:
````diff
@@ -94,8 +109,8 @@ page.
 ```javascript
-CD.getCORS('http://example.com/page', function(data, status) {
-  CD.$('h1')[0].innerHTML = data.header;
+CD.getCORS('http://example.com/page').then((response) {
````
I think this needs a fat arrow `=>`

snackzone commented on tests/tests.js:
```diff
@@ -12,81 +26,81 @@ CD._searchString = function() {
-  var self = this;
+const imageSuccessStub = function() {
+  const self = this;
   setTimeout(function() {
```
Couldn't you just use an ES6 arrow function instead of defining `self`?

istateside replied:
Unfortunately not, because these two functions are declared at the top of the test file, absent of any scope. There's no clear way to capture the scope inside the imageSuccessStub besides grabbing the context at the moment. We could do it with a closure, but that looks messier than just capturing the `this` value.

if these functions were defined on an actual object or in a class, arrow functions would work

snackzone commented on src/cropduster.js:
```diff
@@ -1,189 +1,183 @@
+    if (!doc) {
+      doc = document;
+    }
     return Array.prototype.slice.call(doc.querySelectorAll(selector));
```
this can just be `[...doc.querySelectorAll(selector)]`

snackzone commented on src/cropduster.js:
```diff
@@ -1,189 +1,183 @@
-  suspend: function(maxSuspension, msg) {
-    msg = msg || "manual suspension";
+  pause(maxSuspension, msg) {
+    msg = msg || 'manual suspension';
```
you can just move this into the function's default value for this argument: `pause(maxSuspension, msg = 'manual suspension') {`

snackzone commented on src/cropduster.js:
```diff
@@ -1,189 +1,183 @@
-    CD._readyToCapture = false;
-    if (typeof(MICapture) == 'undefined') {
-      CD.log("suspended: " + msg);
+    if (typeof MICapture === 'undefined') {
```
This pattern is used a lot, would it make sense to extract it into its own method?

istateside commented:
Responded to all of @snackzone 's feedback- Removing support for CD.suspend/capture is a pretty big deal, although users can use `pause` and `resume` in exactly the same way. The only reason we had to break out two new functions is because suspend/capture is no longer going to work along with the new Capturama changes. From an SE perspective, pause/resume will work identically to suspend/capture. I've added some stuff in the README explaining the new methods, but we should probably find a way to communicate it to them.

I'm unsure what the exact path forward for this is- I spoke to @mnutt about what the plan is for cropduster going forward, and we might stop offering future versions of cropduster as a browser package. Ideally, studio apps will start compiling their own javascript files with dependencies outside of the HTML. If that's the case, then we wouldn't offer cropduster from a script tag pointing to projects.movableink.com anymore, and would only support including cropduster as an NPM dependency.

right now, the webpack config here is building two javascript bundles, spitting out `cropduster.js` and `cropduster.browser.js`. If we want to move forward with this faster, we could just put cropduster.browser.js somewhere, but if the only immediate need is for Studio Utility to use the new cropduster (so that `executeDataSource` returns a Promise), then we could merge this PR, and not bother to put the cropduster.browser.js file anywhere. Pointing studio-utility's package.json to the git repo will download the correct file when we `yarn install`.

snackzone commented:
Readme updates look great!

istateside commented:
Gonna merge this and hold off on putting the compiled files up on projects.movableink.com, so we can move forward with the studio-utility changes.
